### PR TITLE
narrow typeguards in the negative

### LINKED ISSUE #507: negative `TypeGuard`s
```py
def f(x: object) -> TypeGuard[int]:
    return isinstance(x, int)

a: int | str
if f(a):
    reveal_type(a)  # int
else:
    reveal_type(a)  # str
```

### LINKED ISSUE #511: `TypeGuard` usages should complain about non overlapping types
```py
def f(x: object) -> TypeGuard[int]: ...
a: str
assert f(a)  # expected error
```

### LINKED ISSUE #513: warn on always true conditions
```py
if 1: ...  # expect error
assert 1  # expect error
```

### LINKED ISSUE #555: show an unreachable error on an impossible type-guard
```py
from __future__ import annotations
def f(x: object) -> x is list[int]: return True

a: list[object]
assert f(a)
b: str = a  # expect an error here
```

### LINKED ISSUE #558: more typeguard stuff detected
`meet.is_overlapping_types`
```py
    if isinstance(left, TypeGuardedType) or isinstance(  # type: ignore[misc]
        right, TypeGuardedType
    ):
        # A type guard forces the new type even if it doesn't overlap the old.
        return True
```

### LINKED ISSUE #576: a positive typeguard type
That only narrows when true

```py
def is_positive_int(x: int | None):
    return isinstance(x, int) and x > 0
i: int
if is_positive_int(i):
    reveal_type(i)  # int
else:
    reveal_type(i)  # int | None
```

Such examples are all over the place, for example, in the mypy codebase.

The question is, how are these denoted:
- use a conditional expression:
    ```py
    def guard(x: object) -> x is int if True else False: ...
    ````
    The only issue I see with this is that the `True` part is ambiguous and implicitly tied to the context, maybe a `Return[True]` specialform or something would be better.
- `TrueTypeGuard(x=int)` might look a little out of place, but I don't have any issue.

### PR DESCRIPTION
**Pull request by KotlinIsland**
- resolves #576
- resolves #555
- resolves #558
- resolves #511
- resolves #507
- resolves #513

### CODE REVIEW

File: mypy/binder.py [hunk 1]
```diff
@@ -40,7 +40,12 @@ class Frame:
     def __init__(self, id: int, conditional_frame: bool = False) -> None:
         self.id = id
         self.types: dict[Key, Type] = {}
-        self.unreachable = False
+        self.unreachable = 0
+        """
+        0 = reachable
+        1 = unreachable
+        2 = unreachable due to artificial always-true scenarios
```
Comment (DetachHead) [inline mypy/binder.py:47]: enum?

Comment (DetachHead) [reply mypy/binder.py:47]: also document that this is just to work around an existing issue

File: mypy/binder.py [hunk 2]
```diff
@@ -150,8 +155,8 @@ def put(self, expr: Expression, typ: Type) -> None:
             self._add_dependencies(key)
         self._put(key, typ)
 
-    def unreachable(self) -> None:
-        self.frames[-1].unreachable = True
+    def unreachable(self, magic=False) -> None:
```
Comment (DetachHead) [inline mypy/binder.py:158]: artificial?

File: mypy/build.py [hunk 3]
```diff
@@ -1427,9 +1427,9 @@ def find_cache_meta(id: str, path: str, manager: BuildManager) -> CacheMeta | No
     # (all of these attributes can be `None`)
     if (
         m.id != id
-        or m.mtime is None  # type: ignore[redundant-expr]
-        or m.size is None  # type: ignore[redundant-expr]
-        or m.dependencies is None  # type: ignore[redundant-expr]
+        or m.mtime is None
+        or m.size is None
+        or m.dependencies is None
```
Comment (DetachHead) [inline mypy/build.py:1432]: remove the comments about the type ignores

File: mypy/checker.py [hunk 4]
```diff
@@ -611,7 +614,12 @@ def accept_loop(
         with self.binder.frame_context(can_skip=False, conditional_frame=True):
             while True:
                 with self.binder.frame_context(can_skip=True, break_frame=2, continue_frame=1):
-                    self.accept(body)
+                    allow_redundant_expr = self.allow_redundant_expr
+                    self.allow_redundant_expr = True
```
Comment (DetachHead) [inline mypy/checker.py:618]: make sure this is only for while loops

File: test-data/unit/check-based-conditional-types.test [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+[case testConditionalTypePast]
+def f1(x: object) -> "x is int" if True else False: ...  # E: You need to put quotes around the entire type-guard, or enable `__future__.annotations`  [misc]
+def f2(x: object) -> "x is int if True else False": ...
+
+
+[case testConditionalTypeUnsupported]
+from __future__ import annotations
+
+def f1(x: object) -> x is int if int else False: ...  # E: Invalid type comment or annotation  [valid-type] \
```
Comment (DetachHead) [inline test-data/unit/check-based-conditional-types.test:18]: test for `x is int if int else Literal[False]`

File: test-data/unit/check-based-conditional-types.test [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+                                                      # N: The condition can only be "True"
+def f2(x: object) -> int if True else False: ...  # E: Invalid type comment or annotation  [valid-type] \
+                                                  # N: The true branch can only be a type-guard
+def f3(x: object) -> x is int if True else x is str: ...  # E: Invalid type comment or annotation  [valid-type] \
```
Comment (DetachHead) [inline test-data/unit/check-based-conditional-types.test:22]: raise an issue to support this

File: test-data/unit/check-based-typeguard.test
Comment (DetachHead) [inline test-data/unit/check-based-typeguard.test:1]: are multiple typeguards on the same function supported? 
- `a is int & b is str` / `a is int | b is str` - what do both of those mean?
- `a is int & a is str` / `a is int | a is str` - should this be considered valid? what does it mean
- how does this new syntax work with it? `a is int & b is str if True else a is str & b is int`. no brackets makes this pretty unreadable and ambiguous

File: test-data/unit/check-typeguard.test [hunk 7]
```diff
@@ -81,7 +81,7 @@ from typing_extensions import TypeGuard
 def is_str_list(a: List[object]) -> TypeGuard[List[str]]: pass
 def main(a: List[object]):
     if is_str_list(a):
-        reveal_type(a)  # N: Revealed type is "<nothing>"
+        reveal_type(a)
```
Comment (DetachHead) [inline test-data/unit/check-typeguard.test:84]: whats going on here

Comment (KotlinIsland) [reply test-data/unit/check-typeguard.test:84]: now it's unreachable

File: test-data/unit/check-based-typeguard.test
Comment (KotlinIsland) [reply test-data/unit/check-based-typeguard.test:1]: #579

File: test-data/unit/check-based-conditional-types.test (hunk 6 above)
Comment (KotlinIsland) [reply test-data/unit/check-based-conditional-types.test:25]: #580

File: test-data/unit/check-based-typeguard.test
Comment (DetachHead) [reply test-data/unit/check-based-typeguard.test:1]: do we know for sure that this syntax wont cause issues later on when adding support for this new stuff?

Comment (KotlinIsland) [reply test-data/unit/check-based-typeguard.test:1]: I can't foresee any compatibility issues, and if we need to change the implicit return type, we can add a `Returns` later.

**Review by DetachHead — APPROVED** [review]

**Pull request merged**